llvm/include/llvm/CodeGen/GlobalISel/GIMatchTableExecutorImpl.h

```diff
@@ -1485,15 +1485,28 @@ bool GIMatchTableExecutor::executeMatchTable(
       propagateFlags();
       return true;
     }
-    case GIR_MakeTempReg: {
+    case GIR_MakeGenericTempReg: {
       uint64_t TempRegID = readULEB();
       int TypeID = readS8();
 
       State.TempRegisters[TempRegID] =
           MRI.createGenericVirtualRegister(getTypeFromIdx(TypeID));
-      DEBUG_WITH_TYPE(TgtExecutor::getName(),
-                      dbgs() << CurrentIdx << ": TempRegs[" << TempRegID
-                             << "] = GIR_MakeTempReg(" << TypeID << ")\n");
+      DEBUG_WITH_TYPE(TgtExecutor::getName(), {
+        dbgs() << CurrentIdx << ": TempRegs[" << TempRegID
+               << "] = GIR_MakeGenericTempReg(" << TypeID << ")\n";
+      });
+      break;
+    }
+    case GIR_MakeVirtualTempReg: {
+      uint64_t TempRegID = readULEB();
+
+      Register Reg = MRI.createIncompleteVirtualRegister();
```

Comment:
Where does the class get set here? I'd expect any create action to carry the class at least

Reply:
The class is set by `constrainOperandRegClass()` called when executing a subsequent `GIR_ConstrainOperandRC` or `GIR_[Root]ConstrainSelectedInstOperands` action. This function doesn't seem to query a virtual register's type, but it calls a few TII / TRI hooks (e.g. `TRI.getConstrainedRegClassForOperand()`) that might in theory do that.

Reply:
In general that would need the type. I think it's a bad idea to let this be so far removed, and tablegen needs to supply the class to use

Reply:
> tablegen needs to supply the class to use

This doesn't seem possible ATM. Producer and consumer of a virtual register might not agree on register class, in which case `constrainOperandRegClass` would split it into two virtual registers and insert a cross-class copy. We would need to lift this machinery to tablegen level. I'm not sure this is what we want or if it's even possible?

Reply:
If we want to provide the register class for temporaries right when creating them, it will need larger refactoring.

Specifically, we will need to provide the desired register class(-es) to node importing functions, or create temporaries in advance and pass them to these functions. This is necessary for leaf nodes that don't know the resulting register class, and may also be useful for instruction nodes so that we can generate a cross-class copy(-es) if necessary.

Reply:
The temporary would be the class of the producer. An additional copy is only needed if getCommonSubClass for the producer and consumer can't find anything. I guess we could handle that in the selector implementation without having tablegen figure that out, not sure if having tablegen figure that out would be better

Reply:
No code can make a sound decision based on an incomplete register, and this doesn't work for instructions that do not have concrete register class constraints. Even ones with register class constraints may need a type to decide what class they should be using

Reply:
I'm confused. 

> An additional copy is only needed if getCommonSubClass for the producer and consumer can't find anything. I guess we could handle that in the selector implementation

If we delay emitting this copy, then the register's class would be invalid for the consumer and we shouldn't make decisions based on invalid register class either.

> this doesn't work for instructions that do not have concrete register class constraints

This is new to me. How can instruction's register operand not have a concrete register class?

Reply:
To be clear I am 100% against passing around incomplete virtual registers. Incomplete virtual registers are an implementation detail of virtual registers that are not part of the contract of valid MIR that code should have to reason about. 

> This is new to me. How can instruction's register operand not have a concrete register class?

This is all generic instructions like COPY and IMPLICIT_DEF, but targets can do the same with their own pseudos. This is part of why the tablegen forms of EXTRACT_SUBREG need an explicit register class operand

Reply:
> To be clear I am 100% against passing around incomplete virtual registers. Incomplete virtual registers are an implementation detail of virtual registers that are not part of the contract of valid MIR that code should have to reason about.

I see your point and I can agree with it.

> If we delay emitting this copy, then the register's class would be invalid for the consumer and we shouldn't make decisions based on invalid register class either.

I'm not comfortable with creating registers with a class that is only valid for the producer. Why do you think fixing the register class by emitting a COPY is better left to instruction selector?

Reply:
> Why do you think fixing the register class by emitting a COPY is better left to instruction selector?

I don't, ideally tablegen would fully understand the set of instructions that need to be emitted here and explicitly handle it. But as far as a hack to make progress goes, I think that's better than using incomplete registers if it's hard to get tablegen into a state where that is possible

Reply:
I'll try to do that at tablegen level. The changes are going to be non-local and I'll probably need to split them into several smaller patches; I'll keep this one as a draft.